### ISSUE TITLE
Type stability issue in julia 1.0

### ISSUE THREAD

**Issue by MaximeBouton** [JuliaPOMDP/POMDPs.jl#279]
The `gen` function is type unstable for POMDP objects, below is a MWE, I cannot reproduce it for MDPs from POMDPModels.

```julia
using POMDPs, POMDPModels, Test, Random
@inferred gen(DDNOut(:sp, :r), TigerPOMDP(), true, 1, MersenneTwister(1))
```

**Comment by lassepe** [comment]
I don't think we can do much about this with the current design decisions. The only thing we could try is test it with a Base value type instead of `DDNOut` because `Base.Val` seems to use some [special macro magic](https://github.com/JuliaLang/julia/blob/v1.0.5/base/essentials.jl#L683) to help with type inference but I don't think that this will fix our problem.

The best thing might be to simply point out that for Julia <= v1.0.5 they should use an older POMDPs.jl release and officially discontinue support for 1.0 in the `Project.toml` so that the user gets a warning when running the wrong version.

**Comment by rejuvyesh** [comment]
I don't think we need to discontinue support for v1.0. Some type instability is fine. What's the actual runtime hit?

**Comment by lassepe** [comment]
On my machine, I get the following benchmark results for the `gen` call posted above:

**julia 1.0.5**

```
  memory estimate:  32 bytes
  allocs estimate:  1
  --------------
  minimum time:     31.956 μs (0.00% GC)
  median time:      32.392 μs (0.00% GC)
  mean time:        33.088 μs (0.00% GC)
  maximum time:     368.936 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
---
```

**julia 1.2.0**
```
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     10.103 ns (0.00% GC)
  median time:      10.921 ns (0.00% GC)
  mean time:        11.063 ns (0.00% GC)
  maximum time:     57.123 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999
```

So this a slowdown of about factor 300. As these things run in inner loops, the impact on performance will be quite significant. By discontinuing support for Julia 1.0 it will be still runnable (even though things will be essential non-usable with this slowdown). However, the user will see a warning when `using POMDPs` which seems appropriate for this performance gap.

**Comment by zsunberg** [comment]
Thanks for benchmarking it @lassepe . I wonder if removing support for 1.0 will act like a vote for the julia developers to make a new LTS release.

**Comment by zsunberg** [comment]
I guess my vote is to remove support for 1.0

**Comment by zsunberg** [comment]
Planning on removing support for Julia 1.0 in POMDPs.jl 0.9 for this reason.

**Comment by zsunberg** [comment]
This is fixed in 09a6d76
